odl/trafos/wavelet.py

```diff
@@ -834,14 +957,511 @@ def adjoint(self):
         if self.is_orthogonal:
             return self.inverse
         else:
-            # TODO: put adjoint here
-            return super().adjoint
+            if not isinstance(self.wbasis, pywt.Wavelet):
+                return InverseAdjBiorthWaveletTransform(dom=self.range,
+                                                        nscales=self.nscales,
+                                                        wbasis=self.wbasis)
+            else:
+                return super().adjoint
 
     @property
     def inverse(self):
         """The inverse wavelet transform."""
-        return WaveletTransform(dom=self.range, nscales=self.nscales,
-                                wbasis=self.wbasis, mode=self.mode)
+        if not isinstance(self.wbasis, pywt.Wavelet):
+            return BiorthWaveletTransform(dom=self.range, ran=self.domain,
+                                          nscales=self.nscales,
+                                          wbasis=self.wbasis)
+        elif isinstance(self.wbasis, pywt.Wavelet):
+            return WaveletTransform(dom=self.range, nscales=self.nscales,
+                                    wbasis=self.wbasis, mode=self.mode)
+
+
+class BiorthWaveletTransform(Operator):
```

Comment:
How about adding this to the standard `WaveletTransform` and just raise an error if the filter length is not supported? The init patterns are quite similar, so no worries there. You only have to make `mode` optional (which would be good anyway), expect `wbasis` to be a string, and we're all good I guess.